### PR TITLE
LIB: CLI: add bind_info, bind and unbind functions

### PR DESCRIPTION
**Pull request by glimchb**

### CODE REVIEW

**Comment by lsgunth** [comment]
Thanks, this largely looks great but it needs to be rebased off of master seeing the first commit is already in master. I also have a number of minor nits with the code which I'll comment on in the patch.

I'd also really like to see a better commit description describing what bind/unbind do.

File: cli/main.c [hunk 1]
```diff
@@ -780,6 +780,108 @@ static int temp(int argc, char **argv)
...
+		int phy_port;
+	} cfg = {};
+	const struct argconfig_options opts[] = {
+		DEVICE_OPTION,
+		{"physical", 'f', "", CFG_INT, &cfg.phy_port, required_argument,
+			"physical port number"},
+		{NULL}};
+
+
+
```
Comment (lsgunth) [inline cli/main.c]: Please don't use multiple blank lines. One is sufficient. (This follows Linux Kernel coding convention). There are a couple cases below of this too please fix them all.

File: cli/main.c [hunk 2]
```diff
@@ -780,6 +780,108 @@ static int temp(int argc, char **argv)
...
+	ret = switchtec_bind_info(cfg.dev, &bind_status, cfg.phy_port);
+
+	if (ret < 0) {
+		printf("bind_info_error: %f", ret);
+		return 1;
+	}
+
+	printf("bind state %u\n", bind_status.bind_state);
+	printf("physical port %u binded to %u, partition %u\n", bind_status.phys_port_id,
```
Comment (lsgunth) [inline cli/main.c]: The past tense of "bind" should be "bound" not "binded"

File: cli/main.c [hunk 3]
```diff
@@ -780,6 +780,108 @@ static int temp(int argc, char **argv)
...
+			bind_status.log_port_id, bind_status.par_id);
+
+
+	return 0;
+}
+
+
+static int port_bind(int argc, char **argv)
+{
+	const char *desc = "Bind switchtech PCI2PCI logical port to physical port";
```
Comment (lsgunth) [inline cli/main.c]: s/switchtech/switchtec/ 

Also what is meant by PCI2PCI?

File: cli/main.c [hunk 4]
```diff
@@ -780,6 +780,108 @@ static int temp(int argc, char **argv)
...
+		return 1;
+	}
+
+	return 0;
+}
+
+
+static int port_unbind(int argc, char **argv)
+{
+	const char *desc = "Unbind switchtech PCI2PCI logical port from physical port";
```
Comment (lsgunth) [inline cli/main.c]: Same as above

File: cli/main.c [hunk 5]
```diff
@@ -1651,6 +1753,9 @@ static const struct cmd commands[] = {
 	CMD(log_dump, "Dump firmware log to a file"),
 	CMD(test, "Test if switchtec interface is working"),
 	CMD(temp, "Return the switchtec die temperature"),
+	CMD(port_bind_info, "Return binding info for a physical port"),
+	CMD(port_bind, "Binds switchtec logical and physical ports"),
```
Comment (lsgunth) [inline cli/main.c]: s/Binds/Bind/

File: cli/main.c [hunk 6]
```diff
@@ -1651,6 +1753,9 @@ static const struct cmd commands[] = {
...
+	CMD(port_unbind, "Unbinds switchtec logical port from physical port"),
```
Comment (lsgunth) [inline cli/main.c]: s/Unbinds/Unbind/

File: inc/switchtec/registers.h [hunk 7]
```diff
@@ -406,4 +406,35 @@ struct switchtec_gas {
 	struct pff_csr_regs pff_csr[SWITCHTEC_MAX_PFF_CSR];
 };
 
+struct switchtec_bind_status_in {
```
Comment (lsgunth) [inline inc/switchtec/registers.h]: These definitions do not belong in registers.h as they are not GAS registers.

Either put them in switchtec.c along with the functions that use them or if you think there's value in exposing them to the user put them in their own bind.h similar to pmon.h

File: inc/switchtec/switchtec.h [hunk 8]
```diff
@@ -440,7 +440,11 @@ int switchtec_fw_img_write_hdr(int fd, struct switchtec_fw_footer *ftr,
 int switchtec_fw_is_boot_ro(struct switchtec_dev *dev);
 int switchtec_fw_set_boot_ro(struct switchtec_dev *dev,
 			     enum switchtec_fw_ro ro);
+int switchtec_bind_info(struct switchtec_dev *dev, struct switchtec_bind_status_out *bind_status, int phy_port);
```
Comment (lsgunth) [inline inc/switchtec/switchtec.h]: Max line length is 80 chars. These lines are too long and should be wrapped and indented just like the other functions in this file. (Per Kernel Style)

File: lib/switchtec.c [hunk 9]
```diff
@@ -552,4 +552,64 @@ float switchtec_die_temp(struct switchtec_dev *dev)
 	return temp / 100.;
 }
 
+
+int switchtec_bind_info(struct switchtec_dev *dev, struct switchtec_bind_status_out *bind_status, int phy_port )
```
Comment (lsgunth) [inline lib/switchtec.c]: Line is too long

File: lib/switchtec.c [hunk 10]
```diff
@@ -552,4 +552,64 @@ float switchtec_die_temp(struct switchtec_dev *dev)
...
+{
+	int ret;
+
+	struct switchtec_bind_status_in sub_cmd_id =
+	{ .sub_cmd = MRPC_PORT_INFO,
+	  .phys_port_id = phy_port
```
Comment (lsgunth) [inline lib/switchtec.c]: That open brace { should be on the previous line. Please follow the style in the rest of the code.

File: lib/switchtec.c [hunk 11]
```diff
@@ -552,4 +552,64 @@ float switchtec_die_temp(struct switchtec_dev *dev)
...
+	return 0;
+}
+
+int switchtec_bind(struct switchtec_dev *dev, int par_id, int log_port, int phy_port)
+{
+	int ret;
+	uint32_t output;
+
+	struct switchtec_bind_in sub_cmd_id =
+	{ .sub_cmd = MRPC_PORT_BIND,
```
Comment (lsgunth) [inline lib/switchtec.c]: Same as above

**Review by lsgunth — CHANGES_REQUESTED** [review]
Largely this looks good and I'd love to see it merged but I'd really appreciate it if you can fix all the style issues.

I also noticed you have set the execute bits (100644->100755) on a couple of files. Please do not do that and remove that change from your patch.

Thanks

Logan

**Comment by glimchb** [comment]
Thanks for the comments. Will be working on fixing them

--
Best regards, Boris
Let me google that for you<http://lmgtfy.com/>

**Comment by glimchb** [comment]
done

**Comment by lsgunth** [comment]
Many thanks. I cleaned up a couple more minor style commits and merged it into Microsemi/devel.

Logan

**Pull request closed without merging**